### PR TITLE
Add support for corporate WeChat webhooks

### PR DESCRIPTION
**Pull request by lengyuqu**
这是一个企业微信的webhook 功能
1，提供了企业微信的webhook推送。
这对中国的工作者是有帮助的。这个功能我已经在我的公司使用了一年，由于中国的网络问题，一直没办法进行提交，希望可以获得合并。谢谢

This is a WebHook function of a company WeChat
1. Provide WebHook push for corporate WeChat.
This is helpful for Chinese workers. This feature I have been in my company for a year, because China's network problems have not been submitted, I hope I can get mergers. Thank you

### CODE REVIEW

**Comment by lengyuqu** [comment]
这是一个企业微信的webhook 功能
1，提供了企业微信的webhook推送。
这对中国的工作者是有帮助的。这个功能我已经在我的公司使用了一年，由于中国的网络问题，一直没办法进行提交，希望可以获得合并。谢谢

This is a WebHook function of a company WeChat

Provide WebHook push for corporate WeChat.
This is helpful for Chinese workers. This feature I have been in my company for a year, because China's network problems have not been submitted, I hope I can get mergers. Thank you

File: docs/content/doc/features/webhooks.en-us.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ All event pushes are POST requests. The methods currently supported are:
 - Telegram
 - Microsoft Teams
 - Feishu
+- Wecahtwork
```
Comment (delvh) [inline docs/content/doc/features/webhooks.en-us.md]: Small typo here.

File: options/locale/locale_zh-CN.ini [hunk 2]
```diff
@@ -1747,6 +1747,7 @@ settings.add_telegram_hook_desc=将 <a href="%s">Telegram</a> 集成到您的仓
 settings.add_matrix_hook_desc=将 <a href="%s">Matrix</a> 集成到您的仓库中。
 settings.add_msteams_hook_desc=将 <a href="%s">Microsoft Teams</a> 集成到您的仓库中。
 settings.add_feishu_hook_desc=将 <a href="%s">Feishu</a> 集成到您的仓库中。
+settings.add_wechatwork_hook_desc=将 <a href="%s">Wechatwork</a> 集成到您的仓库中。
```
Comment (a1012112796) [inline options/locale/locale_zh-CN.ini:1851]: these translation files are maintained on crowdin, Just add this new line in ``local_en-US.ini`` is enough, Thanks.

File: services/webhook/wechatwork.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
```
Comment (a1012112796) [inline services/webhook/wechatwork.go]:
```suggestion
// Copyright 2021 The Gitea Authors. All rights reserved.
```

File: templates/admin/hook_new.tmpl [hunk 4]
```diff
@@ -32,6 +32,8 @@
 					<img width="26" height="26" src="{{AssetUrlPrefix}}/img/feishu.png">
 				{{else if eq .HookType "matrix"}}
 					<img width="26" height="26" src="{{AssetUrlPrefix}}/img/matrix.svg">
+				{{else if eq .HookType "wechatwork"}}
+          <img width="26" height="26" src="{{AssetUrlPrefix}}/img/wechatwork.png">
 				{{end}}
```
Comment (a1012112796) [inline templates/admin/hook_new.tmpl]:
```suggestion
				{{else if eq .HookType "wechatwork"}}
                                           <img width="26" height="26" src="{{AssetUrlPrefix}}/img/wechatwork.png">
				{{end}}
```

File: options/locale/locale_zh-CN.ini (hunk 2 above)
Comment (lengyuqu) [reply options/locale/locale_zh-CN.ini:1851]: I have made modifications

File: docs/content/doc/features/webhooks.en-us.md (hunk 1 above)
Comment (lengyuqu) [reply docs/content/doc/features/webhooks.en-us.md]: Has been modified

**Comment by a1012112796** [comment]
Please fix lint
```
5	templates/admin/hook_new.tmpl:
30s
6		36: Wrong indentation type(spaces instead of tabs)
30s
7	templates/org/settings/hook_new.tmpl:
30s
8		31: Wrong indentation type(spaces instead of tabs)
30s
9	templates/repo/settings/webhook/base_list.tmpl:
30s
10		35: Wrong indentation type(spaces instead of tabs)
30s
11		36: Wrong indentation type(spaces instead of tabs)
30s
12	templates/repo/settings/webhook/new.tmpl:
30s
13		29: Wrong indentation type(spaces instead of tabs)
```

```
	services/webhook/wechatwork.go:61:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
162s
4	func (f *WechatworkPayload) Create(p *api.CreatePayload) (api.Payloader, error) {
162s
5	^
162s
6	services/webhook/wechatwork.go:70:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
162s
7	func (f *WechatworkPayload) Delete(p *api.DeletePayload) (api.Payloader, error) {
162s
8	^
162s
9	services/webhook/wechatwork.go:81:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
162s
10	func (f *WechatworkPayload) Fork(p *api.ForkPayload) (api.Payloader, error) {
162s
11	^
162s
12	services/webhook/wechatwork.go:105:3: S1021: should merge variable declaration with assignment on next line (gosimple)
162s
13			var message string
162s
14			^
162s
15	services/webhook/wechatwork.go:146:2: S1021: should merge variable declaration with assignment on next line (gosimple)
162s
16		var pr string
162s
17		^
162s
18	make: *** [Makefile:762: golangci-lint] Error 1
```

File: routers/routes/web.go [hunk 5]
```diff
@@ -503,6 +504,9 @@ func RegisterRoutes(m *web.Route) {
 			m.Post("/matrix/new", bindIgnErr(forms.NewMatrixHookForm{}), repo.MatrixHooksNewPost)
 			m.Post("/msteams/new", bindIgnErr(forms.NewMSTeamsHookForm{}), repo.MSTeamsHooksNewPost)
 			m.Post("/feishu/new", bindIgnErr(forms.NewFeishuHookForm{}), repo.FeishuHooksNewPost)
+			m.Post("/feishu/{id}", bindIgnErr(forms.NewFeishuHookForm{}), repo.FeishuHooksEditPost)
```
Comment (lunny) [inline routers/routes/web.go]: duplicated line

File: services/webhook/wechatwork.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+package webhook
+
+import (
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/git"
+	api "code.gitea.io/gitea/modules/structs"
+	"encoding/json"
+	"fmt"
+	"strings"
```
Comment (lunny) [inline services/webhook/wechatwork.go:10]: Wrong import sequence

**Comment by lengyuqu** [comment]
> 请修理皮棉
> 
> ```
> 5	templates/admin/hook_new.tmpl:
> 30s
> 6		36: Wrong indentation type(spaces instead of tabs)
> 30s
> 7	templates/org/settings/hook_new.tmpl:
> 30s
> 8		31: Wrong indentation type(spaces instead of tabs)
> 30s
> 9	templates/repo/settings/webhook/base_list.tmpl:
> 30s
> 10		35: Wrong indentation type(spaces instead of tabs)
> 30s
> 11		36: Wrong indentation type(spaces instead of tabs)
> 30s
> 12	templates/repo/settings/webhook/new.tmpl:
> 30s
> 13		29: Wrong indentation type(spaces instead of tabs)
> ```
> 
> ```
> 	services/webhook/wechatwork.go:61:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 4	func (f *WechatworkPayload) Create(p *api.CreatePayload) (api.Payloader, error) {
> 162s
> 5	^
> 162s
> 6	services/webhook/wechatwork.go:70:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 7	func (f *WechatworkPayload) Delete(p *api.DeletePayload) (api.Payloader, error) {
> 162s
> 8	^
> 162s
> 9	services/webhook/wechatwork.go:81:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 10	func (f *WechatworkPayload) Fork(p *api.ForkPayload) (api.Payloader, error) {
> 162s
> 11	^
> 162s
> 12	services/webhook/wechatwork.go:105:3: S1021: should merge variable declaration with assignment on next line (gosimple)
> 162s
> 13			var message string
> 162s
> 14			^
> 162s
> 15	services/webhook/wechatwork.go:146:2: S1021: should merge variable declaration with assignment on next line (gosimple)
> 162s
> 16		var pr string
> 162s
> 17		^
> 162s
> 18	make: *** [Makefile:762: golangci-lint] Error 1
> ```

It has been fixed

File: options/locale/locale_cs-CZ.ini [hunk 7]
```diff
@@ -1701,6 +1701,7 @@ settings.add_telegram_hook_desc=Integrovat <a href="%s">Telegram</a> do vašeho
 settings.add_matrix_hook_desc=Integrovat <a href="%s">Matrix</a> do vašeho repozitáře.
 settings.add_msteams_hook_desc=Integrovat <a href="%s">Microsoft Teams</a> do vašeho repozitáře.
 settings.add_feishu_hook_desc=Integrovat <a href="%s">Feishu</a> do vašeho repozitáře.
+settings.add_wechatwork_hook_desc=Integrovat <a href="%s">Wechatwork</a> do vašeho repozitáře.
```
Comment (lunny) [inline options/locale/locale_cs-CZ.ini:1761]: Don't add translation directly. Only en-US should be changed from the PR

Comment (lengyuqu) [reply options/locale/locale_cs-CZ.ini:1761]: Thank you, I am the first submission code.

**Comment by lunny** [comment]
@lengyuqu Please resolve the conflicts and ensure the CI PASS.

**Comment by lengyuqu** [comment]
> @lengyuqu请解决冲突并确保 CI PASS。

fixed

File: docs/content/doc/features/webhooks.zh-tw.md [hunk 8]
```diff
@@ -26,6 +26,7 @@ Gitea 的儲存庫事件支援 web hook。這可以有儲存庫管理員在設
 - Telegram
 - Microsoft Teams
 - Feishu
+- Wechatwork
```
Comment (lunny) [inline docs/content/doc/features/webhooks.zh-tw.md:29]: Could you also add this to all other webhooks.*.md

File: services/webhook/wechatwork.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+			return nil, err
+		}
+
+		title = fmt.Sprintf("[%s] Pull request review %s : #%d %s", p.Repository.FullName, action, p.Index, p.PullRequest.Title)
+		text = p.Review.Content
+
+	}
+
+	return newWechatworkMarkdownPayload("# " + title + "\r\n\r\n >" + text), nil
+
```
Comment (lunny) [inline services/webhook/wechatwork.go:160]: unnecessary blank line.

**Comment by lunny** [comment]
@lengyuqu some slight changes, otherwise LGTM.

**Review by lengyuqu — COMMENTED** [review]
A file translation

**Comment by lengyuqu** [comment]
> 请修复棉绒
> 
> ```
> 5	templates/admin/hook_new.tmpl:
> 30s
> 6		36: Wrong indentation type(spaces instead of tabs)
> 30s
> 7	templates/org/settings/hook_new.tmpl:
> 30s
> 8		31: Wrong indentation type(spaces instead of tabs)
> 30s
> 9	templates/repo/settings/webhook/base_list.tmpl:
> 30s
> 10		35: Wrong indentation type(spaces instead of tabs)
> 30s
> 11		36: Wrong indentation type(spaces instead of tabs)
> 30s
> 12	templates/repo/settings/webhook/new.tmpl:
> 30s
> 13		29: Wrong indentation type(spaces instead of tabs)
> ```
> 
> ```
> 	services/webhook/wechatwork.go:61:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 4	func (f *WechatworkPayload) Create(p *api.CreatePayload) (api.Payloader, error) {
> 162s
> 5	^
> 162s
> 6	services/webhook/wechatwork.go:70:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 7	func (f *WechatworkPayload) Delete(p *api.DeletePayload) (api.Payloader, error) {
> 162s
> 8	^
> 162s
> 9	services/webhook/wechatwork.go:81:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 10	func (f *WechatworkPayload) Fork(p *api.ForkPayload) (api.Payloader, error) {
> 162s
> 11	^
> 162s
> 12	services/webhook/wechatwork.go:105:3: S1021: should merge variable declaration with assignment on next line (gosimple)
> 162s
> 13			var message string
> 162s
> 14			^
> 162s
> 15	services/webhook/wechatwork.go:146:2: S1021: should merge variable declaration with assignment on next line (gosimple)
> 162s
> 16		var pr string
> 162s
> 17		^
> 162s
> 18	make: *** [Makefile:762: golangci-lint] Error 1
> ```
already fixed

**Comment by lunny** [comment]
@lengyuqu Please resolve the conflicts and I think this could be merged after another L-G-T-M .

**Comment by lunny** [comment]
@lengyuqu lint failed.

**Comment by lengyuqu** [comment]
> 请修复棉绒
> 
> ```
> 5	templates/admin/hook_new.tmpl:
> 30s
> 6		36: Wrong indentation type(spaces instead of tabs)
> 30s
> 7	templates/org/settings/hook_new.tmpl:
> 30s
> 8		31: Wrong indentation type(spaces instead of tabs)
> 30s
> 9	templates/repo/settings/webhook/base_list.tmpl:
> 30s
> 10		35: Wrong indentation type(spaces instead of tabs)
> 30s
> 11		36: Wrong indentation type(spaces instead of tabs)
> 30s
> 12	templates/repo/settings/webhook/new.tmpl:
> 30s
> 13		29: Wrong indentation type(spaces instead of tabs)
> ```
> 
> ```
> 	services/webhook/wechatwork.go:61:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 4	func (f *WechatworkPayload) Create(p *api.CreatePayload) (api.Payloader, error) {
> 162s
> 5	^
> 162s
> 6	services/webhook/wechatwork.go:70:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 7	func (f *WechatworkPayload) Delete(p *api.DeletePayload) (api.Payloader, error) {
> 162s
> 8	^
> 162s
> 9	services/webhook/wechatwork.go:81:1: receiver name f should be consistent with previous receiver name p for WechatworkPayload (golint)
> 162s
> 10	func (f *WechatworkPayload) Fork(p *api.ForkPayload) (api.Payloader, error) {
> 162s
> 11	^
> 162s
> 12	services/webhook/wechatwork.go:105:3: S1021: should merge variable declaration with assignment on next line (gosimple)
> 162s
> 13			var message string
> 162s
> 14			^
> 162s
> 15	services/webhook/wechatwork.go:146:2: S1021: should merge variable declaration with assignment on next line (gosimple)
> 162s
> 16		var pr string
> 162s
> 17		^
> 162s
> 18	make: *** [Makefile:762: golangci-lint] Error 1
> ```

**Comment by lengyuqu** [comment]
> @lenyuqu lint 失败。

I'm fixing this problem.

**Review by lunny — APPROVED** [review]

File: docs/content/doc/features/webhooks.zh-cn.md [hunk 10]
```diff
@@ -15,4 +15,17 @@ menu:
 
 # Webhooks
 
+
```
Comment (6543) [inline docs/content/doc/features/webhooks.zh-cn.md]:
```suggestion
```

File: docs/content/doc/features/webhooks.zh-cn.md [hunk 11]
```diff
@@ -15,4 +15,17 @@ menu:
...
+- Gitea (也可以是 GET 請求)
+- Gogs
+- Slack
+- Discord
+- Dingtalk
+- Telegram
+- Microsoft Teams
+- Feishu
+- Wechatwork
```
Comment (6543) [inline docs/content/doc/features/webhooks.zh-cn.md:29]:
```suggestion
- Wechatwork

```

**Review by 6543 — APPROVED** [review]

**Pull request merged**